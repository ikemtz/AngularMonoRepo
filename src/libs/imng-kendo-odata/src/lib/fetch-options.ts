export interface FetchOptions {
  /**
   * Collection of child table properties that are rendered on a grid
   * This is important because the OOB kendo will not query child tables correctly
   */
  boundChildTableProperties?: BoundChildTableProperty[];
  /**
   * Collection of property names that are of type: nullable Date
   */
  dateNullableProps?: string[];
  /**
   * Collection of property names that are of type: nullable UTC DateTime
   *    Note: these properties will be converted to local date time
   */
  utcNullableProps?: string[];
  /**
   * Set to true to force request via cache-busting.
   * Default: false;
   */
  bustCache?: boolean;
  /**
   * Additional parameters to be added to the request.
   * This is useful for passing additional query parameters that are not part of the OData state.
   */
  additionalParams?: Record<string, string>;
}

export interface BoundChildTableProperty {
  /**
   * This is the child table name
   */
  table: string;
  /**
   * This is the field name on the child table that is bound
   */
  field: string;
  /**
   * This controls if 'all' child records have to meet the the conditions
   * or if 'any' of the child records have to meet the condition.
   * Default: 'any'
   */
  linqOperation?: 'all' | 'any';
}
