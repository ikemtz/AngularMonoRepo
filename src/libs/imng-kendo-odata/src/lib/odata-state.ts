import { State, FilterDescriptor, CompositeFilterDescriptor, SortDescriptor } from '@progress/kendo-data-query';
import { IdType } from 'imng-nrsrx-client-utils';

export interface ODataState extends State {
  expanders?: Array<Expander | string>;
  selectors?: string[];
  inFilters?: InFilter[];
  childFilters?: CompositeChildFilterDescriptor;
  count?: boolean;
  transformations?: string;
  compute?: Array<Computation | string>;
}
export interface Computation {
  /** This can also be a static value */
  fieldA: string | number;
  /** This can also be a static value */
  fieldB: string | number;
  operator: 'mul' | 'div' | 'add' | 'sub' | 'mod' | string;
  /** This MUST differ from the names of declared or dynamic properties of the identified resources. */
  alias: string;
}

export interface InFilter {
  /** This value will default to 'and' in cases where there are additional filters specified. */
  logic?: 'or' | 'and';
  field: string;
  values: IdType[];
}

export interface CompositeChildFilterDescriptor extends CompositeFilterDescriptor {
  filters: Array<CompositeChildFilterDescriptor | ChildFilterDescriptor>;
}

export interface ChildFilterDescriptor extends FilterDescriptor {
  /**  This value will default to 'and' in cases where there are additional filters specified. */
  childTableNavigationProperty: string;
  linqOperation: 'all' | 'any';
}

export interface Expander {
  table: string;
  selectors?: string[];
  expander?: string | Expander;
  filter?: CompositeFilterDescriptor;
  sort?: Array<SortDescriptor>;
}
export function isExpander(source: string | Expander): source is Expander {
  return !!(source as Expander)?.table;
}
export function isComputation(source: string | Computation): source is Computation {
  return !!(source as Computation)?.operator;
}
