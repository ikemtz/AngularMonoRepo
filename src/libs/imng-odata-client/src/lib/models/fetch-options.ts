export interface FetchOptions {
  /**
   * Collection of child table properties that are rendered on a grid
   */
  boundChildTableProperties?: BoundChildTableProperty[];
  /**
   * Collection of property names that are of type: nullable Date
   */
  dateNullableProps?: string[];
  /**
   * Collection of property names that are of type: nullable UTC DateTime
   *    Note: these properties will be converted to local date time
   */
  utcNullableProps?: string[];
  /**
   * Set to true to force request via cache-busting.
   * Default: false;
   */
  bustCache?: boolean;
}

export interface BoundChildTableProperty {
  /**
   * This is the child table name
   */
  table: string;
  /**
   * This is the field name on the child table that is bound
   */
  field: string;
  /**
   * This controls if 'all' child records have to meet the the conditions
   * or if 'any' of the child records have to meet the condition.
   * Default: 'any'
   */
  linqOperation?: 'all' | 'any';
}
