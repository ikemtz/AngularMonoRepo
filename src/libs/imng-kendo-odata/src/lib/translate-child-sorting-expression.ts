import { SortDescriptor } from '@progress/kendo-data-query';
import { BoundChildTableProperty } from './fetch-options';
import { Expander, isExpander, ODataState } from './odata-state';
import { distinct } from 'imng-nrsrx-client-utils';

export function translateChildSortingExpression(
  odataState: ODataState,
  childTableProperties: BoundChildTableProperty[],
): ODataState {
  if (!childTableProperties || childTableProperties.length === 0) {
    return odataState;
  }
  const childTableStrings: string[] = distinct(childTableProperties.map((x) => x.table));
  const childTablePropertyStrings: string[] = childTableProperties.map((x) => `${x.table}.${x.field}`);
  const filterPredicate = (x: SortDescriptor) => childTablePropertyStrings.indexOf(x.field) > -1;
  const sortedColumns = odataState.sort
    ?.filter(filterPredicate)
    .map((m) => ({ ...m, childProperty: m.field.split('.') }));
  if (sortedColumns?.length > 0) {
    odataState.sort = odataState.sort?.filter((x) => !filterPredicate(x));

    odataState.expanders = odataState.expanders.map((m) => (isExpander(m) ? { ...m } : m));
    const expanders = odataState.expanders
      .filter((t: Expander) => childTableStrings.indexOf(t.table) > -1)
      .map((t: Expander) => {
        t.sort = (t.sort || []).filter(
          (f) => !sortedColumns.find((s) => t.table === s.childProperty[0] && f.field === s.childProperty[1]),
        );
        return t;
      });
    sortedColumns.forEach((x) => {
      const expander = expanders.find((e) => e.table === x.childProperty[0]);
      expander.sort.push({ field: x.childProperty[1], dir: x.dir });
    });
  }
  return odataState;
}
