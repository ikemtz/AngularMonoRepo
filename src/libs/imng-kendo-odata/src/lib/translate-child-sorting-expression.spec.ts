import { BoundChildTableProperty } from './fetch-options';
import { ODataState } from './odata-state';
import { translateChildSortingExpression } from './translate-child-sorting-expression';

describe('translateChildSortingExpression', () => {
  it('should handle child sorting', async () => {
    const odataState: ODataState = { expanders: [{ table: 'a' }], sort: [{ field: 'a.b', dir: 'desc' }] };
    const childTableProperties: BoundChildTableProperty[] = [{ table: 'a', field: 'b' }];
    const result = translateChildSortingExpression(odataState, childTableProperties);
    expect(result).toMatchSnapshot();
  });
  it('should handle child sorting on previously sorted column', async () => {
    const odataState: ODataState = {
      expanders: [{ table: 'a', sort: [{ field: 'b', dir: 'desc' }] }],
      sort: [{ field: 'a.b', dir: 'desc' }],
    };
    const childTableProperties: BoundChildTableProperty[] = [{ table: 'a', field: 'b' }];
    const result = translateChildSortingExpression(odataState, childTableProperties);
    expect(result).toMatchSnapshot();
  });
});
