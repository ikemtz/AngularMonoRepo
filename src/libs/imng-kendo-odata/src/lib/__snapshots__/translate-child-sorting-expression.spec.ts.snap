// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`translateChildSortingExpression should handle child sorting 1`] = `
{
  "expanders": [
    {
      "sort": [
        {
          "dir": "desc",
          "field": "b",
        },
      ],
      "table": "a",
    },
  ],
  "sort": [],
}
`;

exports[`translateChildSortingExpression should handle child sorting on previously sorted column 1`] = `
{
  "expanders": [
    {
      "sort": [
        {
          "dir": "desc",
          "field": "b",
        },
      ],
      "table": "a",
    },
  ],
  "sort": [],
}
`;
